contracts/VeJoeStaking.sol

```diff
@@ -0,0 +1,267 @@
+    VeJoeToken public veJoe;
+
+    /// @notice The maximum ratio of veJOE to staked JOE
+    /// For example, if user has `n` JOE staked, they can own a maximum of `n * maxCap` veJOE.
+    uint256 public maxCap;
+
+    /// @notice Rate of veJOE generated per sec per JOE staked
+    uint256 public baseGenerationRate;
+
+    /// @notice Boosted rate of veJOE generated per sec per JOE staked
+    uint256 public boostedGenerationRate;
+
+    /// @notice Percentage of total staked JOE user has to deposit in order to start
+    /// receiving boosted benefits, in parts per 100.
+    /// @dev Specifically, user has to deposit at least `boostedThreshold/100 * totalStakedJoe` JOE.
+    /// The only exception is the user will also receive boosted benefits if its their first
+    /// time staking.
+    uint256 public boostedThreshold;
+
+    /// @notice The length of time a user receives boosted benefits
+    uint256 public boostedDuration;
+
+    mapping(address => UserInfo) public userInfos;
+
+    event Deposit(address indexed user, uint256 amount);
+    event Withdraw(address indexed user, uint256 amount);
+    event Claim(address indexed user, uint256 amount);
+
+    /// @notice Initialize with needed parameters
+    /// @param _joe Address of the JOE token contract
+    /// @param _veJoe Address of the veJOE token contract
+    /// @param _baseGenerationRate Rate of veJOE generated per sec per JOE staked
+    /// @param _boostedGenerationRate Boosted rate of veJOE generated per sec per JOE staked
+    /// @param _boostedThreshold Percentage of total staked JOE user has to deposit to be boosted
+    /// @param _boostedDuration Length of time a user receives boosted benefits
+    function initialize(
+        IERC20Upgradeable _joe,
+        VeJoeToken _veJoe,
+        uint256 _baseGenerationRate,
+        uint256 _boostedGenerationRate,
+        uint256 _boostedThreshold,
+        uint256 _boostedDuration
+    ) public initializer {
+        require(address(_joe) != address(0), "VeJoeStaking: unexpected zero address for _joe");
+        require(address(_veJoe) != address(0), "VeJoeStaking: unexpected zero address for _veJoe");
+        require(
+            _boostedGenerationRate > _baseGenerationRate,
+            "VeJoeStaking: expected _boostedGenerationRate to be greater than _baseGenerationRate"
+        );
+        require(_boostedThreshold <= 100, "VeJoeStaking: expected _boostedThreshold to be less than or equal to 100");
+
+        __Ownable_init();
+
+        maxCap = 100;
+        joe = _joe;
+        veJoe = _veJoe;
+        baseGenerationRate = _baseGenerationRate;
+        boostedGenerationRate = _boostedGenerationRate;
+        boostedThreshold = _boostedThreshold;
+        boostedDuration = _boostedDuration;
+    }
+
+    /// @notice Set maxCap
+    /// @param _maxCap the new maxCap
+    function setMaxCap(uint256 _maxCap) external onlyOwner {
+        require(_maxCap > 0, "VeJoeStaking: expected new _maxCap to be greater than 0");
```

Comment:
What happens in the edge case where someone has more vejoe than the new `maxcap` would allow.

Reply:
We should also have a maximum value (advice is add a few 0's to the maximum you can think off) for `_maxCap`, this is to stop overflow errors.

Reply:
> What happens in the edge case where someone has more vejoe than the new maxcap would allow.

Great question! What would happen currently is they wouldn't be able to claim any more `veJoe` but if the `maxCap` were to increase in the future, they would be able to. We should probably ask murloc/mountain farmer what the desired behavior should be though.

Reply:
> We should also have a maximum value (advice is add a few 0's to the maximum you can think off) for _maxCap, this is to stop overflow errors.

ooo fancy. I made the upper limit `100000` for now but also added a `TODO` to align with murloc/mountain farmer what the desired upper limit should be